### PR TITLE
Subscriptions

### PR DESCRIPTION
**Pull request by leebyron**
I coded up a mostly working version of what I was talking about with @robzhu earlier today to use the [AsyncIterator](https://github.com/tc39/proposal-async-iteration) protocol to handle subscriptions.

This uses the same architectural model as [apollographql/graphql-subscriptions](https://github.com/apollographql/graphql-subscriptions) in that it provides a stateful in-process subscription, and on each "publish" then GraphQL's existing `execute()` function is called.

In the context of AsyncIterator this can be expressed as a "map" operation which I've implemented here and added some tests for.

The actual guts of the implementation are really not interesting at all - almost entirely recycled from functions in `execute.js`. In fact, perhaps going this direction would necessitate a minor refactoring of that file to reduce the obvious code duplication between the two and make this change even smaller.

To produce the pubsub AsyncIterator, a field on the root subscription type can provide a `subscribe` resolver function, or the default resolver function can be used, which allows for providing subscription functions via the rootObject - mirroring the existing pattern used for mutation fields.

Another idea that has come up is to have the ability to provide a global function (not defined in your schema) which is responsible for accepting a parsed GraphQL document and returning a corresponding pubsub AsyncIterator. That's effectively the design of [apollographql/graphql-subscriptions](https://github.com/apollographql/graphql-subscriptions) as it is now. I think this idea is actually just a subscriptions specific variant of the similar request in https://github.com/graphql/graphql-js/issues/733 - allowing a custom "default resolver" to apply. We could consider this idea separately and reap its value for all types in GraphQL, not just subscriptions!

There are some TODOs scattered around here - mostly highlighting places where errors during subscription setup are not being handled yet. What value should a failed subscription setup produce?

### CODE REVIEW

File: src/subscription/__tests__/eventEmitterAsyncIterator.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/**
+ *  Copyright (c) 2017, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
```
Comment (leebyron) [inline src/subscription/__tests__/eventEmitterAsyncIterator.js:10]: This file is only used for tests. It's an example of a mapping between `EventEmitter` and `AsyncIterator`. Especially since graphql-subscriptions uses `EventEmitter` internally now, this file could help bridge the gap between the interfaces.

File: src/subscription/__tests__/subscribe-test.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+      importantEmail: { type: EmailEventType },
+    }
+  });
+
+  const emailSchema = new GraphQLSchema({
+    query: QueryType,
+    subscription: SubscriptionType
+  });
+
+  it('produces a payload per subscription event', async () => {
```
Comment (leebyron) [inline src/subscription/__tests__/subscribe-test.js:80]: Here's a test to see subscriptions in use. Note that above the schema defines very few resolvers. I'm relying on the default resolver to call functions in `rootValue` to keep the example very simple.

File: src/subscription/__tests__/subscribe-test.js [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+          {
+            from: 'joe@graphql.org',
+            subject: 'Hello',
+            message: 'Hello World',
+            unread: false,
+          },
+        ],
+      },
+      importantEmail() {
+        return eventEmitterAsyncIterator(pubsub, 'importantEmail');
```
Comment (leebyron) [inline src/subscription/__tests__/subscribe-test.js:96]: For example, this `data` is the `rootValue`, so when the `importantEmail` field is resolved, it returns an AsyncIterator created from the simple in-memory `EventEmitter` object we're using for this test.

File: src/subscription/__tests__/subscribe-test.js [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    // Another new email arrives, before subscription.next() is called.
+    expect(sendImportantEmail({
+      from: 'hyo@graphql.org',
+      subject: 'Tools',
+      message: 'I <3 making things',
+      unread: true,
+    })).to.equal(true);
+
+    // The next waited on payload will have a value.
+    expect(await subscription.next()).to.deep.equal({
```
Comment (leebyron) [inline src/subscription/__tests__/subscribe-test.js:175]: This is one way to use the AsyncIterable API, but probably not how most will use it. `next()` returns a `Promise<IteratorResult>`, so `await asyncIterator.next()` produces a `IteratorResult` object with the payload. To continue to receive payloads, you have to continue to call `next()`.

A stage-3 TC39 proposal gives us `for await` loop syntax which makes calling `next()` continuously much easier. However to test events happening in a certain order (testing buffering in each direction), I've used the lower level direct API in the test.

In practice, people will use an adapter which converts AsyncIterator to some network protocol like websockets.

Roughly:

```js
websocket.addEventListener('close', () => subscription.return())
try {
  for await (let result of subscription) {
    websocket.send(JSON.stringify(result));
  }
} finally {
  websocket.close(1011);
}
```

File: src/subscription/mapAsyncIterator.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+const ASYNC_ITERATOR_SYMBOL =
+  typeof Symbol === 'function' && Symbol.asyncIterator || '@@asyncIterator';
+
+/**
+ * Given an AsyncIterator and a callback function, return a new AsyncIterator
+ * which produces values mapped via calling the callback function.
+ */
+export default function mapAsyncIterator<T, U>(
```
Comment (leebyron) [inline src/subscription/mapAsyncIterator.js]: This utility function is critical for a simple implementation for `subscribe()`

File: src/subscription/subscribe.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    payload => execute(
+      schema,
+      document,
+      payload,
+      contextValue,
+      variableValues,
+      operationName
+    )
+  );
+}
```
Comment (leebyron) [inline src/subscription/subscribe.js:86]: Everything above line 80 in this function is a straight copy from `execute()`.

This function is responsible for:

1. Given a GraphQL document, produce `AsyncIterable<Payload>` (where resolvers get to decide what type `Payload` is)
2. Define a mapping function `Payload => ExecutionResult` (just call `execute()`)
3. Map the AsyncIterable over this function to produce `AsyncIterable<ExecutionResult>`

File: src/subscription/subscribe.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+  // TODO: resolveFn could throw!
+  const subscription = resolveFn(rootValue, args, context, info);
+
+  invariant(
+    isIterable(subscription),
+    'Subscription must return async-iterator.'
+  );
+
+  return subscription;
+}
```
Comment (leebyron) [inline src/subscription/subscribe.js]: The large majority of this function is recycled from `executeOperation()` and `resolveField()`. There's almost nothing new here.

The differences: 

* It's only resolving a single field instead of an object of fields.
* It will call `subscribe` instead of calling `resolve` if defined on the schema
* A couple extra invariants

**Review by leebyron — COMMENTED** [review]
I've added some inline comments to guide review

File: src/subscription/__tests__/mapAsyncIterator-test.js [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+      await doubles.next()
+    ).to.deep.equal({ value: 2, done: false });
+    expect(
+      await doubles.next()
+    ).to.deep.equal({ value: 4, done: false });
+
+    // Throw error
+    expect(
+      await doubles.throw('ouch')
+    ).to.deep.equal({ value: 'ouchouch', done: false });
```
Comment (robzhu) [inline src/subscription/__tests__/mapAsyncIterator-test.js:183]: This doesn't seem intuitive. Are you sure you want the map function to apply to error handling?

File: src/subscription/subscribe.js [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+ *
+ * Returns an AsyncIterator
+ *
+ * If the arguments to this function do not result in a legal execution context,
+ * a GraphQLError will be thrown immediately explaining the invalid input.
+ */
+export function subscribe(
+  schema: GraphQLSchema,
+  document: DocumentNode,
+  rootValue?: mixed,
```
Comment (robzhu) [inline src/subscription/subscribe.js:49]: What is the purpose of the rootValue parameter here?

File: src/subscription/subscribe.js [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+      document,
+      payload,
+      contextValue,
+      variableValues,
+      operationName
+    )
+  );
+}
+
+function resolveSubscription(
```
Comment (robzhu) [inline src/subscription/subscribe.js:88]: Given the overloaded use of "resolve", I suggest naming this "buildSubscription".

File: src/subscription/subscribe.js (hunk 9 above)
Comment (leebyron) [reply src/subscription/subscribe.js:49]: Same purpose as in `execute` - to provide the top level value context for performing the operation.

In practice a common use of this is to store root field resolvers, and the test case leverages this ability by placing the resolver function for `importantEmail` on the rootValue

File: src/subscription/__tests__/mapAsyncIterator-test.js (hunk 8 above)
Comment (leebyron) [reply src/subscription/__tests__/mapAsyncIterator-test.js:183]: This test needs more comments. The map function doesn't apply to error handling - it only applies to yielded values. This test is checking an edge case where if the parent generator catches an error and continues that the continued yielded values are also mapped

File: src/subscription/subscribe.js (hunk 10 above)
Comment (leebyron) [reply src/subscription/subscribe.js:88]: Sure - we can get to naming concerns. I actually think `resolve` is the correct name in that it's the result of the Resolve algorithm - but perhaps this would be more clear after some refactoring. Since this function is almost entirely recycling of code written in `execute.js`, I'd love for those parts to be more clearly referenced by function call so the purpose and name of this function is more clear.

File: src/subscription/subscribe.js [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+  const fieldNode = fieldNodes[0];
+  const fieldName = fieldNode.name.value;
+  const fieldDef = getFieldDef(exeContext.schema, type, fieldName);
+  invariant(
+    fieldDef,
+    'This subscription is not defined by the schema.'
+  );
+
+  // TODO: make GraphQLSubscription flow type special to support defining these?
+  const resolveFn = (fieldDef: any).subscribe || defaultFieldResolver;
```
Comment (robzhu) [inline src/subscription/subscribe.js:116]: I think we should make "subscribe" support the override scenario rather than the default. I suspect most subscriptions will use the same resolve logic (for example, using the subscription name). In that case, we would want to make defaultFieldResolver injectable. Thoughts?

File: src/subscription/subscribe.js [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+  invariant(
+    isIterable(subscription),
+    'Subscription must return async-iterator.'
+  );
+
+  return subscription;
+}
+
+function isIterable(value) {
```
Comment (robzhu) [inline src/subscription/subscribe.js]: Should this logic more closely mirror how you define ASYNC_ITERATOR_SYMBOL above?

File: src/subscription/subscribe.js (hunk 11 above)
Comment (leebyron) [reply src/subscription/subscribe.js:116]: Yeah! I mentioned this in the 2nd to last paragraph in the top description. I definitely agree - I think making the defaultFieldResolver injectable overall would be valuable thing to do. It doesn't need to be specific to subscriptions. That way we reduce behavior divergence between the two.

File: src/subscription/subscribe.js (hunk 12 above)
Comment (leebyron) [reply src/subscription/subscribe.js]: Yes, I think you're right. The more accurate test would be to looking for a property named with that symbol.

File: src/subscription/__tests__/subscribe-test.js [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+      unread: true,
+    })).to.equal(false); // No more listeners.
+
+    // Awaiting a subscription after closing it results in completed results.
+    expect(await subscription.next()).to.deep.equal({
+      done: true,
+      value: undefined,
+    });
+  });
+
```
Comment (robzhu) [inline src/subscription/__tests__/subscribe-test.js:213]: For most subscriptions, the AsyncIterable stream is infinite (take for example, likes on a post, there's no theoretical "final like"). Rather, we need to fluently express the case of unsubscribing.

File: src/subscription/__tests__/subscribe-test.js (hunk 4 above)
Comment (robzhu) [reply src/subscription/__tests__/subscribe-test.js:175]: A few examples using ```for await ``` would be helpful. As an aside, is there anything special about our set of babel plugin transforms? Could we switch to the ES2017 preset?

File: src/subscription/__tests__/subscribe-test.js (hunk 13 above)
Comment (leebyron) [reply src/subscription/__tests__/subscribe-test.js:213]: That's right the vast majority will be infinite. 

For expressing unsubscription, AsyncIterable models that as the subscriber `return`ing the AsyncIterable. In syntactical async generator functions (like one managing a Redis connection) that would trigger a jump to a `finally` block where code would live to close the stream.

However there are edge cases where subscriptions could complete - both due to error (system failure, or maybe you hit a rate limit?) and product based (you're subscribed to updates from another logged in user who just logged off?)

File: src/subscription/__tests__/subscribe-test.js (hunk 4 above)
Comment (leebyron) [reply src/subscription/__tests__/subscribe-test.js:175]: Our set is more selective. It's essentially ES2017 minus things known to have serious runtime burden. There are also a few transforms we use "loose" mode for to tradeoff deviation from spec-correctness in favor of better runtime perf

**Comment by leebyron** [comment]
Added a rev which does some minor refactoring of execute.js to help better illustrate how subscribe.js should be almost entirely recycling of existing functions

File: src/subscription/subscribe.js (hunk 6 above)
Comment (leebyron) [reply src/subscription/subscribe.js:86]: The PR has been updated to remove the copied code

**Review by robzhu — APPROVED** [review]
Accepting so that we can move ahead with building a few examples of unsubscription.

**Comment by Urigo** [comment]
This is great!

I my opinion you got the exact thing that should sit inside `graphql.js` and left out the things that people can get opinionated about and shouldn’t be generalized.

That means that once this will be released, there will still be a need for a `subscriptionManager` like we have inside `graphql-subscriptions`.
Inside our current `subscriptionManager` we can now gladly remove all the GraphQL parsing, validating and just use the new `subscribe` method.
That would simply our implementation a lot.
Most people will still need a subscriptionManager in order to map channels to subscriptions, filter results before and after execution, get the transport to subscribe to that subscriptionManager to send the processes results to the clients only after all those things and to handle authentication.
People would use Apollo’s subscriptionManager or write their own.

I believe a lot of the mapping and filtering can probably be performed by the Pub/Sub engine itself, so I’m happy you excluded that from the PR. people can use the Apollo implementation of that to easily to all of that or just to understand what they need to do in order to implement it themselves (adding any logic they want for throttling, batching, etc…)

So I think we should still recommend to not use this implementation alone. for example, to connect the transport to the subscription manager and not directly to the subscribe interface.

On a similar note, we are taking Rob’s advice and adding a warning on our in-memory Pub/Sub engine that this should not be used in production.

A question - AsyncIterator is a great interface for us! I love it! very much resembles Observables and brings a lot of it’s benefits.  just from curiosity, why did you chose that and not observables?  I know observables are sensitive subject and also we have no idea what will happen with them in future, I'm more asking for my personal knowledge, if there are technical reasons I should learn about.
* Going in that direction, what are your thought on returning async iterator on queries as well?  That together with the resolvers being able to return asyncIterators that might open the door for community members to start working on some implementation of @live. Or maybe you could “subscribe” to a query?!?
Something we are doing this week to unblock this (but not the main reason we are doing this..) is to change our `ws-transport` to support sending any operation types - queries, mutations and subscriptions, over WebSocket.
